test/lms_test.c

```diff
@@ -0,0 +1,538 @@
+{
+    int ret = 0;
+    LMS_ACVP_TEST_DATA *td = &lms_testdata[1];
+    EVP_PKEY *pkey = NULL;
+    EVP_SIGNATURE *sig = NULL;
+    EVP_PKEY_CTX *ctx = NULL;
+    size_t siglen = 0;
+    const int step = 3;
+    unsigned char sigdata[4096];
+
+    if (!TEST_ptr(pkey = lms_pubkey_from_data(td->pub, td->publen))
+            || !TEST_ptr(sig = EVP_SIGNATURE_fetch(libctx, "LMS", NULL))
+            || !TEST_ptr(ctx = EVP_PKEY_CTX_new_from_pkey(libctx, pkey, NULL)))
+        goto end;
+
+    if (!TEST_size_t_le(td->siglen + 16, sizeof(sigdata)))
+        goto end;
+
+    OPENSSL_cleanse(sigdata, sizeof(sigdata));
+    memcpy(sigdata, td->sig, td->siglen);
+
+    ret = 0;
+    for (siglen = 0; siglen < td->siglen + 16; siglen += step) {
+        if (siglen == td->siglen)   /* ignore the size that should pass */
+            continue;
+        if (!TEST_int_eq(EVP_PKEY_verify_message_init(ctx, sig, NULL), 1))
+            goto end;
+        if (!TEST_int_eq(EVP_PKEY_verify(ctx, sigdata, siglen,
+                                         td->msg, td->msglen), 0)) {
+            ret = -1;
+            goto end;
+        }
+    }
+
+    ret = 1;
+end:
+    if (ret == -1)
+        TEST_note("Incorrectly accepted signature key of length"
+                  " %u (expected %u)", (unsigned)siglen, (unsigned)td->siglen);
+    EVP_SIGNATURE_free(sig);
+    EVP_PKEY_CTX_free(ctx);
+    EVP_PKEY_free(pkey);
+
+    return ret == 1;
+}
+
+static int lms_verify_bad_pub_sig_test(void)
+{
+    LMS_ACVP_TEST_DATA *td = &lms_testdata[1];
+    int ret = 0, i = 0;
+    EVP_PKEY *pkey = NULL;
+    EVP_SIGNATURE *sig = NULL;
+    EVP_PKEY_CTX *ctx = NULL;
+    unsigned char *pub = NULL;
+    const int step = 1;
+
+    /* Copy the public key data so that we can corrupt it */
+    if (!TEST_ptr(pub = OPENSSL_memdup(td->pub, td->publen)))
+        return 0;
+
+    if (!TEST_ptr(sig = EVP_SIGNATURE_fetch(libctx, "LMS", NULL)))
+        goto end;
+
+    for (i = 0; i < (int)td->publen; i += step) {
+        if (i > 0) {
+            EVP_PKEY_free(pkey);
+            EVP_PKEY_CTX_free(ctx);
+            pkey = NULL;
+            ctx = NULL;
+            pub[i - step] ^= 1;
+        }
+        pub[3] ^= 1;
+        /* Corrupting the public key may cause the key load to fail */
+        pkey = lms_pubkey_from_data(pub, td->publen);
+        if (pkey == NULL)
+            continue;
+        if (!TEST_ptr(ctx = EVP_PKEY_CTX_new_from_pkey(libctx, pkey, NULL)))
+            continue;
+        if (!TEST_int_eq(EVP_PKEY_verify_message_init(ctx, sig, NULL), 1))
+            continue;
+        /* We expect the verify to fail */
+        if (!TEST_int_eq(EVP_PKEY_verify(ctx, td->sig, td->siglen,
+                                         td->msg, td->msglen), 0)) {
+            ret = -1;
+            goto end;
+        }
+    }
+
+    ret = 1;
+end:
+    if (ret == -1)
+        TEST_note("Incorrectly passed when byte %d of the public key"
+                  " was corrupted", i);
```

Comment:
Since there's only one spot where we set `ret = -1` it might be more obvious to do this there - and why not `TEST_error`?

Reply:
I've relocated this but not changed it to TEST_error since it immediately follows a TEST_int_eq which will produce an error message.